Add browser version compatibility settings (#1704)
mozilla/extension-workshop: src/content/documentation/publish/version-compatibility.md

---
layout: sidebar
title: Firefox version compatibility
permalink: /documentation/publish/version-compatibility/
topic: Publish
tags: [add-ons, intermediate, tutorial, webextensions, amo, distribution]
contributors: [dotproto]
last_updated_by: dotproto
date: 2023-10-05
---

<!-- Page Hero Banner -->

{% capture page_hero_banner_content %}

# Firefox version compatibility

Learn how to customize Firefox version compatibility settings on addons.mozilla.org and when to use this feature.

{% endcapture %}
{% include modules/page-hero.liquid,
    content: page_hero_banner_content
%}

<!-- END: Overview Page Hero Banner -->

<!-- Content with Table of Contents Module -->

{% capture content_with_toc %}

## `browser_specific_settings` vs. AMO compatibility settings

There are two ways to control Firefox version compatibility settings for an add-on: using the `browser_specific_settings` field in the add-on's manifest or with the compatibility controls in AMO's version management interface. Each setting controls a different part of the addon

> Due to the way these two control mechanisms interact, it's possible for a developer to accidentally make an extension available for download in a browser that will fail to install it. To avoid this, we recommend that developes only use `browser_specific_settings` to control browser compatibility settings.

Reviewer: ```suggestion
Due to the way these two control mechanisms interact, it's possible for a developer to accidentally make an extension available for download in a browser that will fail to install it. To avoid this, we recommend that developers only use `browser_specific_settings` to control browser compatibility settings.
```

Version compatibility controls allow you to manage which versions of Firefox are able to install your extension. Developers can use these controls to make an extension available on Android or prevent incompatible versions of Firefox from installing the extension. For example, say you are working on an extension that heavily depends on the Declarative Net Request API. To ensure that incompatible versions of Firefox don't install it, you would want to set the minimum compatible version of Firefox to 113, the first release that supported this API.

There are two ways to control Firefox version compatibility settings for an add-on. Each setting controls a different part of the addon installation experience.

<dl>
<dt>The <code>browser_specific_settings</code> field in manifest.json</dt>
<dd style="padding-left: 2rem;">

This setting controls the version compatibility check that the user's web browser performs during installation. It is also used to populate AMO's compatibility controls.

</dd>

<dt>AMO's compatibility controls</dt>
<dd style="padding-left: 2rem;">

This setting controls what extensions appear in search results on AMO and are available for installation.</dd>

> </dl>

Author: For some reason when I was drafting I really wanted to use a definition list. While I do like the visual rhythm of my approach, I would much rather avoid the inline CSS tweaks I implemented to make it work.

Happy to run with your suggestion on this :)

| `strict_min_version` | Specifies the lowest version of the browser that can load the extension.                                          |
| `strict_max_version` | Specifies the highest version of the browser that can load the extension.<br>*Avoid unless absolutely necessary.* |

### Android compatibility

In order to signal that your extension is compatible with Android, your manifest must at least set `browser_specific_settings.gecko_android` to an empty object. If you don't, AMO will assume that the extension is not compatible with Android and will not list it as available on Android. This can be manually overridden using AMO's compatibility controls. Omitting `"gecko_android"` does not affect Firefox for Android's ability to install the extension.

```json
{
  "name": "Desktop & Android extension",
  "version": "1.0",
  "manifest_version": 2,
  "browser_specific_settings": {
    "gecko_android": {}
  }
}
```

{% endcapture %}
{% include modules/one-column.liquid,
  id: "browser-specific-settings"
  content: content
%}

<!-- END: Content with Table of Contents Module -->

<!-- Single Column Body Module -->

{% capture content %}

## AMO compatibility setting

> AMO's browser version compatibility settings determine which versions of Firefox can search for and install a given extension. The browser compatibility settings for a newly uploaded version of an extension are automatically by the `browser_specific_settings` of the extension's manifest.

Reviewer 2: ```suggestion
AMO's browser version compatibility settings determine whether the extension is listed in a search of AMO from a version of Firefox and, therefore, make it available for installation. When an extension is uploaded to AMO, the extension's browser compatibility settings in AMO are set from the `browser_specific_settings` of the extension's manifest.
```

Author: > whether the extension is listed in a search of AMO from a version of Firefox and, therefore, make it available for installation.

To my knowledge this isn't quite accurate. As worded, this implies that since you can't search for an extension you can't install it. Even without search, a user could navigate directly to an extension's AMO listing and try to install the extension. If the extension is not flagged as supporting Android, AMO will disable the "add to Firefox" button.

> 

Reviewer 2: @dotproto don't we need something here to explain why you would change the settings? I can only think of something along the lines of "For example, if you discover that your extension doesn't work on the latest version of Firefox due to a regression or similar issue in the web extension API, you can limit the maximum version that your extension install is on until the issue is fixed. "

Also, a note along the lines of "Remember that, if you set a wider range of supported versions in AMO, your extension may list in searches but not installed due to the Firefox version limits set in its manifest file."

Author: +1 to both of those suggestions. I'll integrate locally & push and update soon.

> Use the following instructions to manually configure browser compatibility settings.

Reviewer 2: ```suggestion
To manually configure the browser compatibility settings:
```

1. Visit the [Add-on Developer Hub](https://addons.mozilla.org/en-US/developers/). <br/><br/> <img src="/assets/img/publish/configure_versions_dev_hub.png" alt="Screen showing the My Add-ons list in the Add-on Developer Hub" style="box-shadow:0 0 0.5em gray;" />

> 2. In My Add-ons, find the add-on you want to configure and click **Edit Product Page**.

Reviewer 2: ```suggestion
2. In My Add-ons, find the add-on you want to configure and select **Edit Product Page**.
```

> 3. Click **Manage Status & Versions** in the left navigation menu. A list of all your extension versions displays.<br/><br/> <img src="/assets/img/publish/configure_versions_manage_menu.png" alt="Screen showing the all versions list for an extension" style="box-shadow:0 0 0.5em gray;" />

Reviewer 2: ```suggestion
3. Select **Manage Status & Versions** in the left navigation menu. A list of all your extension versions displays.<br/><br/> <img src="/assets/img/publish/configure_versions_manage_menu.png" alt="Screen showing the all versions list for an extension" style="box-shadow:0 0 0.5em gray;" />
```

> 4. Click the version number you want to configure.

Reviewer 2: ```suggestion
4. Choose the extension version you want to configure.
```

5. In the compatibility section, for the product you want to specify versions for, select the earliest and latest product versions this version of your add-on is compatible with. Remember, if you used `browser_specific_settings.gecko_android` the compatibility settings for Android are locked.<br/><br/> <img src="/assets/img/publish/configure_versions_manage_page.png" alt="Screen showing the manage version options, including the settings for compatibility." style="box-shadow:0 0 0.5em gray;" />

> 6. Click **Save Changes** to apply your edits.

Reviewer 2: ```suggestion
6. Select **Save Changes** to apply your edits.
```

---
layout: sidebar

> title: Submitting an add-on

Reviewer 2: ```suggestion
title: Firefox version compatibility
```

permalink: /documentation/publish/version-compatibility/
topic: Publish
tags: [add-ons, intermediate, tutorial, webextensions, amo, distribution]
contributors: [dotproto]
last_updated_by: dotproto
date: 2023-10-05
---

<!-- Page Hero Banner -->

{% capture page_hero_banner_content %}

# Firefox version compatibility

Learn how to customize Firefox version compatibility settings on addons.mozilla.org and when to use this feature.

{% endcapture %}
{% include modules/page-hero.liquid,
    content: page_hero_banner_content
%}

<!-- Content with Table of Contents Module -->

{% capture content_with_toc %}

> ## AMO settings vs. browser_specific_settings

Reviewer 2: ```suggestion
## AMO settings vs. `browser_specific_settings`
```

> ## browser_specific_settings

Reviewer 2: ```suggestion
## Manifest `browser_specific_settings` key
```

> ## AMO compatibility

Reviewer 2: ```suggestion
## AMO compatibility setting
```

You can control the versions of Firefox that can install your add-on by configuring a version's compatibility settings. 

> 1. Visit the [Add-on Developer Hub](https://addons.mozilla.org/en-US/developers/). <br/><br/> <img src="/assets/img/publish/configure_versions_dev_hub.png" style="box-shadow:0 0 0.5em gray;" />

Reviewer 2: ```suggestion
1. Visit the [Add-on Developer Hub](https://addons.mozilla.org/en-US/developers/). <br/><br/> <img src="/assets/img/publish/configure_versions_dev_hub.png" alt="Screen showing the My Add-ons list in the Add-on Developer Hub" style="box-shadow:0 0 0.5em gray;" />
```

> 1. Visit the [Add-on Developer Hub](https://addons.mozilla.org/en-US/developers/). <br/><br/> <img src="/assets/img/publish/configure_versions_dev_hub.png" alt="Screen showing the My Add-ons list in the Add-on Developer Hub" style="box-shadow:0 0 0.5em gray;" />

Author: @rebloor, I seem to recall that in another review you mentioned that you preferred the use of Markdown-style image includes over HTML ones. Is that correct?

The main reason I was leaning towards using HTML was because it allows us to add a box shadow or border to the image, which helps the reader differentiate between screenshot content and article text.

Reviewer 2: The usual best practice is to include the border in the imagerather than switching to HTML. This was how images were presented priginally, I think I'm bummed into a couple of legacy examples recently. However, given that we've gone in this direction and it now encompasses many of the images, I suggest we continue.

Author: That seems reasonable to me. Do you have a preference for how we should handle screenshots on other pages? I'm open to either baking the border into the image or using CSS to decorate images before upload. If we decide to embrace CSS, I'd like to create a [shortcode](https://www.11ty.dev/docs/shortcodes/) or convention in Markdown to make our image handling more consistent.

2. In My Add-ons, find the add-on you want to configure and click **Edit Product Page**.

> 3. Click **Manage Status & Versions** in the left navigation menu. <br/><br/> <img src="/assets/img/publish/configure_versions_manage_menu.png" style="box-shadow:0 0 0.5em gray;" />

Reviewer 2: ```suggestion
3. Click **Manage Status & Versions** in the left navigation menu. A list of all your extension versions displays.<br/><br/> <img src="/assets/img/publish/configure_versions_manage_menu.png" alt="Screen showing the all versions list for an extension" style="box-shadow:0 0 0.5em gray;" />
```

4. Click the version number you want to configure.

> 5. In the compatibility section, for the product you want to specify versions for, select the earliest and latest product versions this version of your add-on is compatible with. <br/><br/> <img src="/assets/img/publish/configure_versions_manage_page.png" style="box-shadow:0 0 0.5em gray;" />

Reviewer 2: Is it worth adding a reminder about the Android limitations?
```suggestion
5. In the compatibility section, for the product you want to specify versions for, select the earliest and latest product versions this version of your add-on is compatible with. Remember, if you used `browser_specific_settings.gecko_android` the compatibility settings for Android are locked.<br/><br/> <img src="/assets/img/publish/configure_versions_manage_page.png" alt="Screen showing the manage version options, including the settings for compatibility." style="box-shadow:0 0 0.5em gray;" />
```
